go/ct/spc/coordination_test.go

```diff
@@ -0,0 +1,186 @@
+	for i := 0; i < e.numStates; i++ {
+		g := generator.Clone()
+		res := consume(g)
+		if res == rlz.ConsumeAbort {
+			return rlz.ConsumeAbort
+		}
+	}
+	return rlz.ConsumeContinue
+}
+
+func (e testCondition) String() string {
+	return "Test Condition"
+}
+
+func TestCoordination_GenerateMultipleStatesPerRule(t *testing.T) {
+	numJobs := runtime.NumCPU()
+	seed := 0
+	var counter atomic.Int64
+	opFunction := func(state *st.State) rlz.ConsumerResult {
+		counter.Add(1)
+		return rlz.ConsumeContinue
+	}
+	printFunction := func(time time.Duration, rate float64, current int64) {}
+
+	statesPerRule := 15
+
+	//var condition testCondition
+	rules := []rlz.Rule{}
+	testRule := rlz.Rule{
+		Name:      "test_rule",
+		Condition: condition(true, statesPerRule),
+		Effect:    rlz.NoEffect(),
+	}
+	rules = append(rules, testRule)
+	err := ForEachState(rules, opFunction, printFunction, numJobs, uint64(seed), false)
+	if err != nil {
+		t.Errorf("Unexpected error during state generation %v", err)
+	}
+
+	if counter.Load() != int64(statesPerRule) {
+		t.Errorf("unexpected number of evaluated states: %d", counter.Load())
+	}
+}
+
+func TestCoordination_AllRulesAreEnumerated(t *testing.T) {
+	numJobs := runtime.NumCPU()
+	seed := 0
+	var counter atomic.Int64
+	opFunction := func(state *st.State) rlz.ConsumerResult {
+		counter.Add(1)
+		return rlz.ConsumeContinue
+	}
+	printFunction := func(time time.Duration, rate float64, current int64) {}
+
+	//var condition testCondition
+	rules := []rlz.Rule{}
+	testRuleFit := rlz.Rule{
+		Name:      "test_rule",
+		Condition: condition(true, 1),
+		Effect:    rlz.NoEffect(),
+	}
+	testRuleNoFit := rlz.Rule{
+		Name:      "test_rule",
+		Condition: condition(false, 1),
+		Effect:    rlz.NoEffect(),
+	}
+
+	for i := 0; i < rand.Intn(42); i++ {
+		rules = append(rules, testRuleFit)
+		rules = append(rules, testRuleNoFit)
+	}
+	numRules := len(rules)
+
+	err := ForEachState(rules, opFunction, printFunction, numJobs, uint64(seed), false)
+	if err != nil {
+		t.Errorf("Unexpected error during state generation %v", err)
+	}
+
+	if counter.Load() != int64(numRules)/2 {
+		t.Errorf("unexpected number of evaluated states: %d", counter.Load())
+	}
+
+	counter.Store(0)
+	err = ForEachState(rules, opFunction, printFunction, numJobs, uint64(seed), true)
+	if err != nil {
+		t.Errorf("Unexpected error during state generation %v", err)
+	}
+
+	if counter.Load() != int64(numRules) {
+		t.Errorf("wrong number of evaluated states in full mode: %d vs. %d", counter.Load(), numRules)
+	}
+}
+
+func TestCoordination_FilterRules(t *testing.T) {
+	filters := []string{"add", "sub", "mul", "copy", "call"}
+	for _, subString := range filters {
+		filter, err := regexp.Compile(subString)
+		if err != nil {
+			t.Error("regular expression not compilable")
+		}
+		rules := Spec.GetRules()
+		rules = FilterRules(rules, filter)
```

Comment:
This test does not check that all rules that should be in the result are actually in the result. It only checks that those that are listed match the criteria.